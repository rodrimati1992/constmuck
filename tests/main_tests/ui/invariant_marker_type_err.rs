// this ensures that the lifetime of the type parameters to `Is*` types are invariant.
// 
// Each function must correspond to exactly one lifetime error,
// currently that error is:
// ```text
// does not necessarily outlive the static lifetime
// ```
use constmuck::{IsContiguous, IsCopy, IsPod, IsTransparentWrapper, IsZeroable, TypeSize};

fn is_contiguous_as<'a, T>(x: IsContiguous<&'a T, u8>) -> IsContiguous<&'static T, u8> {
    x
}
fn is_contiguous_sa<'a, T>(x: IsContiguous<&'static T, u8>) -> IsContiguous<&'a T, u8> {
    x
}

fn is_copy_as<'a, T>(x: IsCopy<&'a T>) -> IsCopy<&'static T> {
    x
}
fn is_copy_sa<'a, T>(x: IsCopy<&'static T>) -> IsCopy<&'a T> {
    x
}

fn is_pod_as<'a, T>(x: IsPod<&'a T>) -> IsPod<&'static T> {
    x
}
fn is_pod_sa<'a, T>(x: IsPod<&'static T>) -> IsPod<&'a T> {
    x
}

fn is_transparent_wrapper_aasa<'a, T>(
    x: IsTransparentWrapper<&'a T, &'a T>,
) -> IsTransparentWrapper<&'static T, &'a T> {
    x
}
fn is_transparent_wrapper_saaa<'a, T>(
    x: IsTransparentWrapper<&'static T, &'a T>,
) -> IsTransparentWrapper<&'a T, &'a T> {
    x
}

fn is_transparent_wrapper_aaas<'a, T>(
    x: IsTransparentWrapper<&'a T, &'a T>,
) -> IsTransparentWrapper<&'a T, &'static T> {
    x
}
fn is_transparent_wrapper_asaa<'a, T>(
    x: IsTransparentWrapper<&'a T, &'static T>,
) -> IsTransparentWrapper<&'a T, &'a T> {
    x
}

fn is_zeroable_as<'a, T>(x: IsZeroable<&'a T>) -> IsZeroable<&'static T> {
    x
}
fn is_zeroable_sa<'a, T>(x: IsZeroable<&'static T>) -> IsZeroable<&'a T> {
    x
}

fn type_size_as<'a, T>(x: TypeSize<&'a T, (), 0>) -> TypeSize<&'static T, (), 0> {
    x
}
fn type_size_sa<'a, T>(x: TypeSize<&'static T, (), 0>) -> TypeSize<&'a T, (), 0> {
    x
}


fn main(){}