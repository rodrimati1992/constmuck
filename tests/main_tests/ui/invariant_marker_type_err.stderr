error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:11:5
   |
11 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsContiguous<&'static T, _>`
              found struct `IsContiguous<&'a T, _>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:10:21
   |
10 | fn is_contiguous_as<'a, T>(x: IsContiguous<&'a T, u8>) -> IsContiguous<&'static T, u8> {
   |                     ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:14:5
   |
14 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsContiguous<&'a T, _>`
              found struct `IsContiguous<&'static T, _>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:13:21
   |
13 | fn is_contiguous_sa<'a, T>(x: IsContiguous<&'static T, u8>) -> IsContiguous<&'a T, u8> {
   |                     ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:18:5
   |
18 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsCopy<&'static T>`
              found struct `IsCopy<&'a T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:17:15
   |
17 | fn is_copy_as<'a, T>(x: IsCopy<&'a T>) -> IsCopy<&'static T> {
   |               ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:21:5
   |
21 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsCopy<&'a T>`
              found struct `IsCopy<&'static T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:20:15
   |
20 | fn is_copy_sa<'a, T>(x: IsCopy<&'static T>) -> IsCopy<&'a T> {
   |               ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:25:5
   |
25 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsPod<&'static T>`
              found struct `IsPod<&'a T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:24:14
   |
24 | fn is_pod_as<'a, T>(x: IsPod<&'a T>) -> IsPod<&'static T> {
   |              ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:28:5
   |
28 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsPod<&'a T>`
              found struct `IsPod<&'static T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:27:14
   |
27 | fn is_pod_sa<'a, T>(x: IsPod<&'static T>) -> IsPod<&'a T> {
   |              ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:34:5
   |
34 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsTransparentWrapper<&'static T, _>`
              found struct `IsTransparentWrapper<&'a T, _>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:31:32
   |
31 | fn is_transparent_wrapper_aasa<'a, T>(
   |                                ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:39:5
   |
39 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsTransparentWrapper<&'a T, _>`
              found struct `IsTransparentWrapper<&'static T, _>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:36:32
   |
36 | fn is_transparent_wrapper_saaa<'a, T>(
   |                                ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:45:5
   |
45 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsTransparentWrapper<_, &'static T>`
              found struct `IsTransparentWrapper<_, &'a T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:42:32
   |
42 | fn is_transparent_wrapper_aaas<'a, T>(
   |                                ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:50:5
   |
50 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsTransparentWrapper<_, &'a T>`
              found struct `IsTransparentWrapper<_, &'static T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:47:32
   |
47 | fn is_transparent_wrapper_asaa<'a, T>(
   |                                ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:54:5
   |
54 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsZeroable<&'static T>`
              found struct `IsZeroable<&'a T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:53:19
   |
53 | fn is_zeroable_as<'a, T>(x: IsZeroable<&'a T>) -> IsZeroable<&'static T> {
   |                   ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:57:5
   |
57 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `IsZeroable<&'a T>`
              found struct `IsZeroable<&'static T>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:56:19
   |
56 | fn is_zeroable_sa<'a, T>(x: IsZeroable<&'static T>) -> IsZeroable<&'a T> {
   |                   ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:61:5
   |
61 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `TypeSize<&'static T, _, _>`
              found struct `TypeSize<&'a T, _, _>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:60:17
   |
60 | fn type_size_as<'a, T>(x: TypeSize<&'a T, (), 0>) -> TypeSize<&'static T, (), 0> {
   |                 ^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
  --> tests/main_tests/ui/invariant_marker_type_err.rs:64:5
   |
64 |     x
   |     ^ lifetime mismatch
   |
   = note: expected struct `TypeSize<&'a T, _, _>`
              found struct `TypeSize<&'static T, _, _>`
note: the lifetime `'a` as defined here...
  --> tests/main_tests/ui/invariant_marker_type_err.rs:63:17
   |
63 | fn type_size_sa<'a, T>(x: TypeSize<&'static T, (), 0>) -> TypeSize<&'a T, (), 0> {
   |                 ^^
   = note: ...does not necessarily outlive the static lifetime
